src/core/allow-bind.c

```diff
@@ -0,0 +1,164 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#if HAVE_LIBBPF
+#include <bpf/bpf.h>
+#endif
+
+#include "allow-bind.h"
+#include "fd-util.h"
+#include "unit.h"
+
+#if BPF_FRAMEWORK
+/* libbpf, clang and llc compile time dependencies are satisfied */
+#include "bpf/allow_bind/allow-bind-skel.h"
+
+static struct allow_bind_bpf *allow_bind_bpf_free(struct allow_bind_bpf *obj) {
+        if (obj && libbpf_get_error(obj) == 0)
+                (void) allow_bind_bpf__destroy(obj);
+
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct allow_bind_bpf *, allow_bind_bpf_free);
+
+static int can_link_bpf_program(struct bpf_program *prog) {
+        _cleanup_(bpf_link_freep) struct bpf_link *link = NULL;
+        int err;
+
+        assert(prog);
+
+        link = bpf_program__attach_cgroup(prog, -1);
+        if (!link)
+                return -ENOMEM;
+
+        err = libbpf_get_error(link);
+        if (err != -EINVAL && err != -EBADF)
+                return err;
+
+        /* EBADF indicates that bpf_link is supported by kernel. */
+        return err == -EBADF;
+}
```

Comment:
maybe: actually:

```c
if (err == -EBADF)
        return true; /* EBADF indicates that bpf_link is supported by kernel. */
if (err == -EINVAL)
        return false;  /* some comment here */
return err;
```

i.e. i think it's better to test for each error type only once fully, instead of filtering first some out, and then checking for them again, if you grok what i mean